requests_oauthlib/oauth2_session.py

```diff
@@ -197,10 +197,18 @@ def fetch_token(self, token_url, code=None, authorization_response=None,
                 redirect_uri=self.redirect_uri, username=username,
                 password=password, **kwargs)
 
-        if (not auth) and username:
-            if password is None:
-                raise ValueError('Username was supplied, but not password.')
-            auth = requests.auth.HTTPBasicAuth(username, password)
+        client_id = kwargs.get('client_id', '')
+        if not auth:
+            if client_id:
+                log.debug('Encoding client_id "%s" with client_secret as Basic auth credentials.', client_id)
+                client_secret = kwargs.get('client_secret', '')
+                client_secret = client_secret if client_secret is not None else ''
```

Comment:
Strictly speaking I think it's not if your goal is to protect from passing `None` to the basic auth function. The `get` with a default is the default used if the key is missing. If someone calls `fetch_token( ... client_secret=None, ... )` then after `client_secret = kwargs.get('client_secret', '')` you'll find that `client_secret` is `None`:

```
>>> client_secret = {'client_secret': None}.get('client_secret', '')
>>> print client_secret
None
>>> client_secret = {}.get('client_secret', '')
>>> print client_secret
>>>```

Unless I'm mistaken about `None` not being possible or not needing to be protected against I think this is best. The other way I was considering was:

```
client_secret = {'client_secret': None}.get('client_secret', '') or ''
```

What do you think?